File: src/main/java/com/github/bzumhagen/sct/GitChangelog.java
package com.github.bzumhagen.sct;

import com.github.zafarkhaja.semver.Version;
import org.eclipse.jgit.api.Git;
import org.eclipse.jgit.api.errors.GitAPIException;
import org.eclipse.jgit.api.errors.NoHeadException;
import org.eclipse.jgit.lib.Repository;
import org.eclipse.jgit.lib.RepositoryBuilder;
import org.eclipse.jgit.revwalk.RevCommit;

import java.io.File;
import java.io.IOException;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.List;
import java.util.Optional;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

class GitChangelog implements Changelog {
    private ChangelogConfiguration config;
    private List<ChangelogChange> changes;
    private File gitDir;

    GitChangelog(ChangelogConfiguration config, File gitDir) {
        this.config = config;
        this.gitDir = gitDir;
        this.changes = buildChanges(gitDir);
    }

    GitChangelog(ChangelogConfiguration config, File gitDir, Version startVersion, Version endVersion) {
        this.config = config;
        this.gitDir = gitDir;
        this.changes = buildChanges(gitDir, startVersion, endVersion);
    }

    private List<ChangelogChange> buildChanges(File gitDir) {
        return buildChanges(gitDir, null, null);
    }

    private List<ChangelogChange> buildChanges(File gitDir, Version startVersion, Version endVersion) {
        List<ChangelogChange> builtChanges = new ArrayList<>();
        try {
            Git git = Git.open(gitDir);
            for(RevCommit commit : git.log().call()) {
                buildChange(commit).map(builtChanges::add);
            }
        } catch (IOException | GitAPIException e) {
            e.printStackTrace();
        }
        return builtChanges;
    }

    private Optional<ChangelogChange> buildChange(RevCommit commit) {
        String commitMessage = commit.getFullMessage();
        String description = commit.getShortMessage();
        Optional<Version> version = getVersionFromMessage(commitMessage);
        Optional<String> tag = getMatchFor(config.getTagPattern(), commitMessage);
        Optional<String> reference = getMatchFor(config.getReferencePattern(), commitMessage);
        if(tag.isPresent() && version.isPresent()) {
            ChangelogChange change = new ChangelogChange(description, version.get(), tag.get(), reference.orElse(""));
            return Optional.of(change);
        } else {
            return Optional.empty();
        }
    }

    private Optional<Version> getVersionFromMessage(String commitMessage) {
        Optional<String> version = getMatchFor(config.getVersionPattern(), commitMessage);
        return version.map(Version::valueOf);
    }

    private Optional<String> getMatchFor(Pattern regex, String data) {
        Matcher matcher = regex.matcher(data);
        if (matcher.find()) {
            return Optional.of(matcher.group(1));
        } else {
            return Optional.empty();
        }
    }

    @Override
    public List<ChangelogChange> getChanges() {
        return this.changes;
    }

    @Override
    public void toMarkdown(Path pathToWriteFile) {
        //TODO
    }
}

File: src/main/scala/com/github/bzumhagen/sct/git/GitChangelog.scala
package com.github.bzumhagen.sct.git

import java.time.{Instant, ZoneId}

import better.files.File
import com.github.bzumhagen.sct._
import com.github.zafarkhaja.semver.Version
import org.eclipse.jgit.api.Git
import org.eclipse.jgit.revwalk.RevCommit
import org.fusesource.scalate._
import org.slf4j.LoggerFactory

import scala.collection.JavaConverters._
import scala.util.matching.Regex

/** A changelog for a git repository.
  *
  *  @constructor create a new git changelog with a config and directory.
  *  @param config changelog configuration
  *  @param gitDir directory containing the git repository
  */
class GitChangelog(val config: ChangelogConfiguration, val gitDir: File) extends Changelog {
  private val logger = LoggerFactory.getLogger(classOf[GitChangelog])

  /** Get all changes from git repository which match the mandatory criteria (i.e. contain version and specified tag) */
  override def getChanges: Seq[ChangelogChange] = {
    val gitRepository = Git.open(gitDir.toJava)
    val gitLog = gitRepository.log().call().asScala

    gitRepository.close()
    gitLog.flatMap(buildChange).toSeq
  }

  /** Generate a file given a file and a non-empty sequence of changes.
    *
    *  @param file file to write markdown into
    *  @param changes sequence of changes to build changelog with
    */
  override def generateChangelog(file: File, changes: Seq[ChangelogChange]): File = {
    require(changes.nonEmpty, "Cannot generate changelog without changes")

    val engine = new TemplateEngine
    val changeBinding = new SmartGroupChangeBinding(getTemplate, changes)
    val defaultBindings = Map(
      "name" -> config.name,
      "showReference" -> config.showReference
    )
    val output = engine.layout(changeBinding.template, defaultBindings ++ changeBinding.buildChangeBindings)
    file.writeText(output)
  }

  private def buildChange(commit: RevCommit): Option[ChangelogChange] = {
    val commitMessage = commit.getFullMessage
    val description = commit.getShortMessage
    val version = getVersionFromMessage(commitMessage)
    val tag = getMatchFor(config.tagPattern, commitMessage)
    val reference = getMatchFor(config.referencePattern, commitMessage).map(ChangelogReference)
    val date = Instant
      .ofEpochSecond(commit.getCommitTime)
      .atZone(ZoneId.systemDefault)
      .toLocalDate

    (version, tag) match {
      case (Some(v), Some(t)) =>
        Some(ChangelogChange(description, v, t, reference, date))
      case (None, Some(_)) =>
        logger.debug(s"Didn't find version in commit message [$commitMessage]")
        None
      case (Some(_), None) =>
        logger.debug(s"Didn't find tag in commit message [$commitMessage]")
        None
      case _ =>
        logger.debug(
          s"Didn't find tag or version in commit message [$commitMessage]")
        None
    }
  }

  private def getVersionFromMessage(commitMessage: String): Option[Version] = {
    val version = getMatchFor(config.versionPattern, commitMessage)
    version.map(Version.valueOf)
  }

  private def getMatchFor(regex: Regex, data: String): Option[String] =
    data match {
      case regex(matchingValue) => Some(matchingValue)
      case _ => None
    }

  private def getTemplate =
    config.templateFile.getOrElse(
      if (config.smartGrouping) {
        "/templates/markdown.mustache"
      } else {
        "/templates/verbose_markdown.mustache"
      })
}
